### PR TITLE
Button

### PR DESCRIPTION
**Pull request by omarhogni**
Added a exe to transform a bool signal into some events for a button.

### CODE REVIEW

File: exes/button/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+struct my_deps {
+  tfc::ipc::bool_signal& long_signal;
+  tfc::ipc::bool_signal& double_signal;
+  tfc::ipc::bool_signal& touch_signal;
+  asio::io_context& ctx;
+  tfc::logger::logger& logger;
+  std::chrono::time_point<system_clock, std::chrono::nanoseconds> initial_time;
+};
+
+// guards
```
Comment (jbbjarnason) [inline exes/button/src/main.cpp:36]: lambdas can be constexpr

File: exes/button/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+struct high {};
+struct low {};
+struct timeout {};
+
+struct my_deps {
+  tfc::ipc::bool_signal& long_signal;
+  tfc::ipc::bool_signal& double_signal;
+  tfc::ipc::bool_signal& touch_signal;
+  asio::io_context& ctx;
+  tfc::logger::logger& logger;
```
Comment (jbbjarnason) [inline exes/button/src/main.cpp:32]: how about owning these references?

Comment (omarhogni) [reply exes/button/src/main.cpp:32]: Sure, that looks better. ctx and logger kept as a reference.

File: exes/button/src/main.cpp (hunk 1 above)
Comment (omarhogni) [reply exes/button/src/main.cpp:36]: I will make them constexpr. Is there anything specific you hope to gain in this case?

Comment (jbbjarnason) [reply exes/button/src/main.cpp:40]: not being globally constructed/destructed, instead being part of the compiled artifact(a symbol). I thought this hsould provoke an clang warning

File: exes/button/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+find_package(glaze CONFIG REQUIRED)
+
+target_link_libraries(button
+  PUBLIC
+    tfc::ipc
+    tfc::base
+    tfc::logger
+    tfc::confman
+    Boost::program_options
+    glaze::glaze
```
Comment (jbbjarnason) [inline exes/button/CMakeLists.txt:20]:
```suggestion
```

File: exes/button/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,38 @@
+add_executable(button src/main.cpp)
+
+find_path(AZMQ_INCLUDE_DIRS "azmq/actor.hpp")
```
Comment (jbbjarnason) [inline exes/button/CMakeLists.txt:3]: azmq is not explicitly used, but sml is used, so maybe change to sml

File: exes/button/src/main.cpp [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+// Actions
+constexpr auto send_signal = [](asio::io_context& ctx, tfc::ipc::bool_signal& signal, tfc::logger::logger& logger) {
+  signal.async_send(true, [&](const std::error_code& err, size_t) {
+    if (err) {
+      logger.warn("Failed to send, {}", err.message());
+    }
+  });
+
+  // Create a timer from a shared_ptr
+  auto timer = std::make_shared<asio::system_timer>(ctx);
```
Comment (jbbjarnason) [inline exes/button/src/main.cpp:83]: my system instead of steady timer?

File: exes/button/CMakeLists.txt (hunk 4 above)
Comment (omarhogni) [reply exes/button/CMakeLists.txt:3]: copy paste error... fixed

File: exes/button/src/main.cpp (hunk 5 above)
Comment (omarhogni) [reply exes/button/src/main.cpp:83]: My thought was that steady timer is not wall time. But we are interacting with the physical world ( someones finger ) but with such large timers I don't think we will find any difference.

The only guarantee steady timer gives is that every call to the timer will return a later point then the last.

Comment (omarhogni) [reply exes/button/src/main.cpp:83]: But then again if the computer applies time corrections during the time I am not sure what the behavior is.

Comment (jbbjarnason) [reply exes/button/src/main.cpp:83]: I think it makes sense to use steady timer for deadline timers. System timer for absolute timers, for example 24.december at 12

Comment (omarhogni) [reply exes/button/src/main.cpp:83]: I think you are correct, it looks like steady_timer is a rename of an older concept called `deadline_timer`. Will fix.

**Pull request merged**